bus/bus.go

```diff
@@ -786,45 +793,34 @@ func New(s Syncer, cm ChainManager, tp TransactionPool, w Wallet, hdb HostDB, cs
 		"GET    /params/download": b.paramsHandlerDownloadGET,
 		"GET    /params/upload":   b.paramsHandlerUploadGET,
 		"GET    /params/gouging":  b.paramsHandlerGougingGET,
-	}), nil
+	}), cleanup, nil
 }
 
-func (b *bus) recordInteractions(interactions []hostdb.Interaction) error {
-	// Check whether someone else is flushing the interactions or if this
-	// thread needs to.
-	var shouldFlush bool
-
-	// No interactionsFlush exists so this thread needs to flush.
+func (b *bus) recordInteractions(interactions []hostdb.Interaction) {
 	b.interactionsMu.Lock()
-	f := b.interactionsFlush
-	if f == nil {
-		f = &interactionsFlush{
-			c: make(chan struct{}),
-		}
-		b.interactionsFlush = f
-		shouldFlush = true
-	}
+	defer b.interactionsMu.Unlock()
+
+	// Append interactions to buffer.
 	b.interactions = append(b.interactions, interactions...)
-	b.interactionsMu.Unlock()
 
-	// If we don't need to flush, we block on the flush channel and then
-	// return the result.
-	if !shouldFlush {
-		<-f.c // TODO: handle shutdown
-		return f.result
+	// If a thread was scheduled to flush the buffer we are done.
+	if b.interactionsFlushTimer != nil {
+		return
 	}
+	// Otherwise we schedule a flush.
+	b.interactionsFlushTimer = time.AfterFunc(b.interactionsFlushInterval, func() {
+		b.flushInteractions()
+	})
+}
 
-	// Otherwise we wait for the interval and apply the interactions.
-	time.Sleep(b.interactionsFlushInterval) // TODO: handle shutdown
-
+func (b *bus) flushInteractions() {
 	b.interactionsMu.Lock()
 	defer b.interactionsMu.Unlock()
-
-	f.result = b.hdb.RecordInteractions(b.interactions)
-	close(f.c)
+	if len(b.interactions) > 0 {
```

Comment:
`nit` Is this ever empty? `flushInteractions` would not be called otherwise? I would personally `return nil` in `RecordInteractions` if it's been given an empty slice.

Reply:
Only during the cleanup if the timing is bad it could be empty. I added a check to `RecordInteractions` as well but this one will prevent the worker to reach out to the bus if there are no interactions which is slightly better.